File: leetcode-274-H-Index/leetcode-274-H-Index.cpp
#include <iostream>
#include <cstdlib>
#include <vector>

int hIndex(std::vector<int>& citations) {
    int max = -1;

    for (const auto& a:citations)
    {
        if(max<a)
        {
            max = a;
        }
    }
    
    std::vector<int> temp_count;
    temp_count.resize(max+1);

    for(const auto& a:citations)
    {
        for(int i = a; 0<=i; --i)
        {
            ++temp_count[i];
        }
    }
    
    for(int i = max; 0<=i; --i)
    {
        if(i<=temp_count[i])
        {
            return i;
        }
    }

    return 0;
}


int main(){
    std::vector<int> vec = {3,0,6,1,5};
    int res = hIndex(vec);

    std::cout << res << std::endl;

    return EXIT_SUCCESS;
}

File: leetcode-274-H-Index/leetcode-274-H-Index.py
class Solution:
    def hIndex(self, citations: list[int]) -> int:
        counters: list[int] = [0] * (max(citations, default=0)+1)
        for c in citations:
            counters[c] += 1
        result: int = 0
        last_val: int = 0
        for i in range((len(counters)-1), -1, -1):
            counters[i] += last_val
            last_val = counters[i]
            if i <= last_val:
                return i
        return result


if __name__ == "__main__":
    citations: list[int] = [3,0,6,1,5]
    result = Solution().hIndex(citations)
    print(result)
